### PR TITLE
FUNC: Notification functionality for sending and receiving chat request

### PR DESCRIPTION
**Pull request by vidushi-rastogi**
Changes Added : 

- API handlers for notification functionality flow
- UI for notifications page
- UI for notifications list in main page
- Logic to send notifications to other users
- Logic for notification actions (accept/reject/remove/cancel).
- Added Tailwind CSS for UI styling
- Added common re-usable UI components

### CODE REVIEW

File: components/layout/header.js [hunk 1]
```diff
@@ -22,7 +22,7 @@ const notifications = (chatNotifications) => (
 )
 
 
-export default function PageHeader({ home, chatNotifications, session, chatPartners, setChatPartners }) {
+export default function PageHeader({ home, notification, chatNotifications, session, chatPartners, setChatPartners }) {
```
Comment (imran1721) [inline components/layout/header.js:25]: this is good 👍

File: components/notifications/notificationsPage.js [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+            <Button danger>Cancel</Button>
+        }
+    </div>
+)
+
+export default function NotificationsPage({ type, notifications }) {
+    console.log('TYPE AND NOTIFICATIONS : ', type, notifications)
+    const totalNotifications = notifications.length;
+    const [data, setData] = useState(notifications.slice(0, 5));
+    console.log('data : ', notifications.slice(0, 5))
```
Comment (imran1721) [inline components/notifications/notificationsPage.js]: remove consoles

File: components/notifications/notificationsPage.js [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+                setData([...data, ...notifications.slice(data.length, data.length + 5)]);
+            }
+            else {
+                setData([...data, ...notifications.slice(data.length, data.length + (totalNotifications - data.length))]);
+            }
+        }
+    };
+    // useEffect(() => {
+    //     appendData();
+    // }, []);
```
Comment (imran1721) [inline components/notifications/notificationsPage.js]: remove commented codes

File: components/notifications/notificationsPage.js [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+            >
+                {(item) => (
+                    <List.Item key={item._id}>
+                        <List.Item.Meta
+                            avatar=
+                            {<Avatar
+                                //src={item.picture.large} 
+                                icon={<UserOutlined />} />}
+                            description={<NotificationDescription type={type} notificationType={item.type} />}
+                            title={type === 'incoming' ? <p>{item.sender}</p> : <p>{item.receiver}</p>}
```
Comment (imran1721) [inline components/notifications/notificationsPage.js:56]: formatting required

File: components/layout/header-components/actionHandlers.js [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+            console.log(`Error: Unable to remove chat request from ${session.user.username} and ${notification.sender}`);
+            popNotification('error', `Unable to reject chat request from ${notification.sender} :(`);
+        }
+    }
+}
+
+export {
+    handleAcceptRequest,
+    handleRemoveRequest
+}
```
Comment (vidushi-rastogi) [inline components/layout/header-components/actionHandlers.js]: new line error

File: components/layout/header-components/actionHandlers.js [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+        const rejectedRequestResponse = await RemoveChatRequest(session.user.username, notification.sender, 'accept', notification._id);
+        const data = await rejectedRequestResponse.json();
+        if (rejectedRequestResponse.status === 200) {
+            setNotifications(data.updatedNotifications.slice(0, 5));
+            let updatedNotifications = userNotifications;
+            updatedNotifications.incomingNotifications = userNotifications?.incomingNotifications?.filter(n => n._id !== notification._id);
+            setUserNotifications(updatedNotifications);
+        }
+        else {
+            console.log(`Error: Unable to remove chat request from ${session.user.username} and ${notification.sender}`);
```
Comment (vidushi-rastogi) [inline components/layout/header-components/actionHandlers.js:97]: are we supposed to log this?

File: components/layout/header-components/notifications.js [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+  setNotifications,
+  userNotifications,
+  setUserNotifications,
+  session,
+  chatPartners,
+  setChatPartners
+}) => {
+  return notifications.length > 0 ?
+  <List
+    size="small"
```
Comment (vidushi-rastogi) [inline components/layout/header-components/notifications.js]: we are not consistent about quotes, somewhere is is single quotes and somewhere its double.

File: components/layout/header-components/notifications.js [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+        <div className="flex content-evenly">
+          <div className="mr-2">
+            <Avatar icon={<BellFilled />} />
+          </div>
+          <div className="mr-3 flex flex-col">
+            <span className="font-bold">@{item.sender} </span>
+            <span className="text-xs"><NotificationDescription type='incoming' notificationType={item.type} /></span>
+          </div>
+          <div>
+            <div className="flex">
```
Comment (vidushi-rastogi) [inline components/layout/header-components/notifications.js]: formatting is needed

File: components/layout/header-components/notifications.js [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+  :
+  <div>
+    <div className="text-center">
+      <ExclamationCircleOutlined className="text-5xl text-blue-500"/>
+    </div>
+    <p>You have viewed all notifications</p>
+  </div>;
+}
+
+const Notifications = ({
```
Comment (vidushi-rastogi) [inline components/layout/header-components/notifications.js:89]: we can move different components to separate files

File: components/layout/header-components/notifications.js [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+            trigger={['click']}
+        >
+            <Badge count={notifications ? notifications?.length : 0}>
+                <BellFilled type="bell" className={styles.headerIcons} />
+            </Badge>
+        </Dropdown>
+    )
+}
+
+export default Notifications;
```
Comment (vidushi-rastogi) [inline components/layout/header-components/notifications.js]: new line error

File: components/layout/header.js [hunk 11]
```diff
@@ -71,31 +68,40 @@ export default function PageHeader({ home, chatNotifications, session, chatPartn
...
-                                    content={notifications(chatNotifications)}
-                                    title="Chat Requests"
-                                    trigger="click"
-                                >
-                                    <BellFilled className={styles.headerIcons} />
-                                </Popover>
-                            </Col>
+                            {!notification ?
+                                <Col span={12}>
+                                    {/* <Popover
```
Comment (vidushi-rastogi) [inline components/layout/header.js]: if this is not needed we can remove.

File: components/notifications/actions/notifyRequestAcception.js [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+          type: 'request_accepted',
+          acknowledgement: false,
+          timestamp: new Date()
+        }),
+        header: {
+          'Content-Type': 'application/json'
+        }
+      })
+    return res;
+}
```
Comment (vidushi-rastogi) [inline components/notifications/actions/notifyRequestAcception.js]: new line error

File: components/notifications/actions/notifyRequestAcception.js [hunk 13]
```diff
@@ -0,0 +1,16 @@
+export const RequestAcceptionNotification = async (currentUser, notification) => {
+    const res = await fetch('/api/postNotification', {
```
Comment (vidushi-rastogi) [inline components/notifications/actions/notifyRequestAcception.js]: return directly, no need of keeping in variable.

File: components/notifications/actions/removeChatRequest.js [hunk 14]
```diff
@@ -0,0 +1,15 @@
+export const RemoveChatRequest = async (receiver, sender, type, notificationId) => {
+    const res = await fetch('/api/deleteChatRequest', {
```
Comment (vidushi-rastogi) [inline components/notifications/actions/removeChatRequest.js]: same as above

File: components/notifications/actions/removeNotification.js [hunk 15]
```diff
@@ -0,0 +1,14 @@
+export const RemoveNotification = async (username, notificationType, notificationId) => {
+    const res = await fetch('/api/deleteNotification', {
```
Comment (vidushi-rastogi) [inline components/notifications/actions/removeNotification.js]: same

File: components/shared-components/backToHome.js [hunk 16]
```diff
@@ -0,0 +1,12 @@
...
+import { ArrowLeftOutlined } from '@ant-design/icons';
+import styles from './shared-components.module.css';
+
+const BackToHome = () => (
+    <Tooltip title='Back'>
+        <ArrowLeftOutlined onClick={() => window.location.href = '/'} className={styles.icons} />
+    </Tooltip>
+)
+
+export default BackToHome;
```
Comment (vidushi-rastogi) [inline components/shared-components/backToHome.js:12]: new line errror

File: components/shared-components/formatDate.js [hunk 17]
```diff
@@ -0,0 +1,10 @@
+import moment from "moment/moment";
+
+const NotificationDateFormat = (dateString) => {
```
Comment (vidushi-rastogi) [inline components/shared-components/formatDate.js]: can be refactored
Also, provide default value

File: components/shared-components/refresh.js [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+
+    return <Tooltip title='Refresh'>
+        <SyncOutlined
+            spin={spin} 
+            onClick={handleRefresh} 
+            className={styles.icons}/>
+    </Tooltip>
+}
+
+export default Refresh;
```
Comment (vidushi-rastogi) [inline components/shared-components/refresh.js]: new line

File: pages/notifications.js [hunk 19]
```diff
@@ -0,0 +1,77 @@
...
+    }, [status, flag, session]);
+
+    return (status === 'authenticated' ?
+        <Layout>
+            <Header>
+                <PageHeader
+                    notification
+                    session={session} />
+            </Header>
+            <div style={{ padding: '0px 10px 0px 10px' }}>
```
Comment (vidushi-rastogi) [inline pages/notifications.js]: use tailwind
file needs formatting

File: components/layout/header-components/actionHandlers.js (hunk 6 above)
Comment (vidushi-rastogi) [reply components/layout/header-components/actionHandlers.js:97]: Yes to know the specific issue causing the error. Can be removed on later stages.

File: components/layout/header-components/notifications.js (hunk 7 above)
Comment (vidushi-rastogi) [reply components/layout/header-components/notifications.js]: Added ES-lint rule to fix the same

**Review by imran1721 — APPROVED** [review]

**Pull request merged**